File: tools/smoke_conhost.js
'use strict';
/* ==========================================================================
   控制台镜像 · 真实浏览器冒烟测试（由 tools/smoke_conhost.py 调起）
   --------------------------------------------------------------------------
   跑法（一般不用手动跑，交给驱动脚本）：
       python tools\smoke_conhost.py

   它用**真实的 Electron（Chromium）**打开虚拟桌面，走一遍用户会走的路：
   开始菜单打开「控制台镜像」→ 列表渲染出真实控制台 → 点一个 →
   内容区真的画出字符栅格（含中文、不重复）→ 只读模式下没有输入区。

   为什么必须做这一步
   ------------------
   前端那几道快速闸门（node --check、跨模块调用名、纯逻辑用例）都查不出
   **「窗口打开了但里面是空的」**。这个项目已经栽过一次：照片网格因为一个
   loading 闸门永远画不出来，而当时接口测试、静态闸门、逻辑用例全绿 ——
   那次正是这个冒烟脚本第一次跑就抓到的。

   环境变量：
       CH_PORT    服务端口
       CH_COOKIE  管理员的会话 Cookie（"fw_session=…"）
       CH_OUT     截图输出目录
       CH_PID     期望在列表里找到的那个控制台进程 pid（确定性目标）
       CH_MARK    期望在内容里读到的标记文本
   ========================================================================== */

const fs = require('fs');
const path = require('path');
const { app, BrowserWindow, session } = require('electron');

app.disableHardwareAcceleration();

const PORT = Number(process.env.CH_PORT);
const COOKIE = String(process.env.CH_COOKIE || '');
const OUT = String(process.env.CH_OUT || '');
const TARGET_PID = String(process.env.CH_PID || '');
const MARK = String(process.env.CH_MARK || '');
const INPUT_MODE = String(process.env.CH_INPUT || '') === '1';
const BASE = 'http://127.0.0.1:' + PORT;

const results = [];
const consoleErrors = [];

function check(label, ok, extra) {
  console.log((ok ? '  [PASS] ' : '  [FAIL] ') + label +
    (extra ? '  —— ' + extra : ''));
  results.push({ label: label, ok: !!ok, extra: extra === undefined ? '' : String(extra) });
}

function wait(ms) {
  return new Promise(function (resolve) { setTimeout(resolve, ms); });
}

async function safeRun(win, code, label) {
  try {
    return await win.webContents.executeJavaScript(code, true);
  } catch (err) {
    console.log('  [WARN] 脚本执行失败（' + (label || '') + '）：' +
      (err && err.message ? err.message : err));
    return { __error: String(err && err.message ? err.message : err) };
  }
}

async function shot(win, name) {
  if (!OUT) {
    return;
  }
  try {
    const image = await win.webContents.capturePage();
    fs.writeFileSync(path.join(OUT, name), image.toPNG());
  } catch (err) {
    console.log('  [WARN] 截图失败：' + err);
  }
}

/** 轮询直到条件为真（从主进程轮询，而不是在页面里挂长 Promise） */
async function waitFor(win, code, timeoutMs, label) {
  const deadline = Date.now() + (timeoutMs || 15000);
  let last = null;
  while (Date.now() < deadline) {
    last = await safeRun(win, code, label);
    if (last === true) {
      return true;
    }
    await wait(350);
  }
  return last;
}

async function main() {
  const eq = COOKIE.indexOf('=');
  if (eq > 0) {
    await session.defaultSession.cookies.set({
      url: BASE + '/',
      name: COOKIE.slice(0, eq),
      value: COOKIE.slice(eq + 1),
      path: '/',
      httpOnly: true
    });
  }

  const win = new BrowserWindow({
    width: 1440,
    height: 900,
    show: true,
    webPreferences: { contextIsolation: true, nodeIntegration: false }
  });

  win.webContents.on('console-message', function (event, level, message) {
    if (level >= 2) {
      consoleErrors.push(String(message).slice(0, 300));
    }
  });
  win.webContents.on('did-fail-load', function (event, code, desc, url) {
    consoleErrors.push('did-fail-load ' + code + ' ' + desc + ' ' + url);
  });

  const run = function (code, label) { return safeRun(win, code, label); };

  await win.loadURL(BASE + '/');
  await wait(2500);

  console.log('=== 1. 桌面 ===');
  check('页面标题正确', String(await run('document.title')).indexOf('远程文件管理') >= 0);

  console.log('=== 2. 开始菜单里应当有「控制台镜像」（管理员才看得到） ===');
  const entry = await run(`(function () {
    var btn = document.getElementById('startBtn');
    if (!btn) { return 'no-start-button'; }
    btn.click();
    var item = document.querySelector('.sm-item[data-action="conhost"]');
    if (!item) {
      var all = Array.prototype.map.call(document.querySelectorAll('.sm-item'),
        function (e) { return e.getAttribute('data-action'); }).join(',');
      return 'no-conhost-entry; 现有: ' + all;
    }
    item.click();
    return 'ok';
  })()`, 'open-conhost');
  check('开始菜单里有「控制台镜像」并点击成功', entry === 'ok', String(entry));

  console.log('=== 3. 窗口打开且**不是空白** ===');
  check('镜像窗口已打开',
    await run("!!document.querySelector('.conhost')", 'has-root'));

  const listed = await waitFor(win,
    "document.querySelectorAll('.conhost-item').length > 0", 20000, 'wait-list');
  const itemCount = await run("document.querySelectorAll('.conhost-item').length");
  check('控制台列表渲染出了条目', listed === true, '条目数=' + itemCount);

  const listInfo = await run(
    "(document.querySelector('.conhost .pv-label') || {}).textContent || ''");
  console.log('      列表摘要：' + listInfo);
  check('列表摘要里报告了控制台数量',
    /共\s*\d+\s*个控制台/.test(String(listInfo)), String(listInfo));
  await shot(win, 'conhost-list.png');

  console.log('=== 4. 选中那个确定性目标（pid=' + TARGET_PID + '） ===');
  /* 用 JSON.stringify 拼进页面脚本：pid 是外部传入的，直接拼字符串容易出
     引号/转义问题（而且模板字符串里再嵌模板字符串会直接把脚本撕开）。 */
  const pidNeedle = JSON.stringify('pid ' + TARGET_PID);
  const picked = await run(`(function () {
    var items = Array.prototype.slice.call(document.querySelectorAll('.conhost-item'));
    var hit = items.filter(function (n) {
      return n.textContent.indexOf(${pidNeedle}) >= 0;
    })[0];
    if (!hit) {
      var all = items.map(function (n) {
        var m = n.querySelector('.conhost-item-meta');
        return m ? m.textContent : '?';
      }).join(' / ');
      return 'not-found; 现有: ' + all;
    }
    hit.click();
    return 'ok';
  })()`, 'select-target');
  check('在列表里找到了目标控制台并点击', picked === 'ok', String(picked));

  console.log('=== 5. 内容区真的画出了字符栅格 ===');
  /* ★ 必须等到**目标那个控制台的标记**出现，不能只等「有行」——
     列表默认会先选中第一个控制台并渲染出内容，所以「有行」在点击之前
     就已经为真了；只等这个条件会立刻通过，然后读到的是**上一个**控制台
     的内容（第一次跑就是这么误判的）。 */
  const markNeedle = JSON.stringify(MARK);
  const gotMark = await waitFor(win,
    "(document.querySelector('.conhost-view') || {}).textContent.indexOf(" +
    markNeedle + ") >= 0", 30000, 'wait-marker');
  const lineCount = await run("document.querySelectorAll('.conhost-line').length");
  check('内容区有渲染出来的行', typeof lineCount === 'number' && lineCount > 0,
    '行数=' + lineCount);

  const text = await run(
    "(document.querySelector('.conhost-view') || {}).textContent || ''", 'view-text');
  console.log('      内容片段：' + String(text).slice(0, 240).replace(/\s+/g, ' '));
  check('内容里包含目标控制台打印的标记「' + MARK + '」', gotMark === true,
    '前 120 字：' + String(text).slice(0, 120).replace(/\s+/g, ' '));
  /* 中文重复是已知坑：全角字符尾格不跳过的话会变成「本本机机」。
     服务端已经处理，这里用「连续两个相同汉字」做粗筛。 */
  check('中文没有出现「每个字重复两遍」',
    !/([\u4e00-\u9fa5])\1/.test(String(text)));

  console.log('=== 6. 输入区的出现与否必须与 allow_input 一致 ===');
  const inputVisible = await run(`(function () {
    var box = document.querySelector('.conhost-input');
    if (!box) { return false; }
    return box.style.display !== 'none' &&
      box.querySelectorAll('input,button').length > 0;
  })()`, 'input-area');
  /* 这一条是**双向**的，比只测一边有价值：
       只读配置下必须没有输入控件（否则界面在骗人 —— 让人以为能打字）；
       可写配置下必须有（否则功能被藏起来了，用户找不到入口）。 */
  check(INPUT_MODE
    ? '开启 allow_input 时应当出现输入区'
    : '未开启 allow_input 时没有输入区（纯只读）',
    inputVisible === INPUT_MODE, 'inputVisible=' + inputVisible);

  await shot(win, 'conhost-content.png');

  console.log('=== 7. 切到「屏幕」视图仍然有内容（且状态栏确认已切换） ===');
  const toggled = await run(`(function () {
    var btns = Array.prototype.slice.call(document.querySelectorAll('.conhost .pv-btn'));
    var b = btns.filter(function (n) { return n.textContent.indexOf('视图') >= 0; })[0];
    if (!b) { return 'no-mode-button'; }
    b.click();
    return b.textContent;
  })()`, 'switch-mode');
  check('找到了视图切换按钮并点击', String(toggled).indexOf('屏幕') >= 0,
    String(toggled));

  /* 等状态栏**自己**说切到屏幕视图了 —— 这比固定 sleep 稳：
     它确认的是「服务端按新模式返回并渲染完成」，而不是「大概过了两秒」。 */
  const screenOk = await waitFor(win,
    "(function () { var s = document.querySelector('.conhost-status');" +
    " return !!s && s.textContent.indexOf('屏幕视图') >= 0; })()",
    25000, 'wait-screen-mode');
  const screenLines = await run("document.querySelectorAll('.conhost-line').length");
  check('屏幕视图已生效且渲染出了行',
    screenOk === true && typeof screenLines === 'number' && screenLines > 0,
    '状态栏确认=' + screenOk + ' 行数=' + screenLines);
  await shot(win, 'conhost-screen.png');

  console.log('=== 8. 控制台错误 ===');
  check('页面控制台没有 error 级日志', consoleErrors.length === 0,
    consoleErrors.slice(0, 6).join(' | '));

  /* ------------------------------------------------------------------
     9. 输入注入（只在 CH_INPUT=1 时跑，因为服务端配置得跟着打开）

     ★ 用 `set /a 111*3` 做验证的技巧：**输入里没有 333，输出里才有**。
       所以「内容里出现 333」这件事只能来自「命令真的被目标 shell 执行了」，
       而不是「把我们自己发出去的字符串又读回来了」—— 后者是自欺欺人的
       假通过（命令回显里带着原始命令行，用普通 echo 就会误判）。
     ------------------------------------------------------------------ */
  if (INPUT_MODE) {
    console.log('=== 9. 输入注入（allow_input = true）===');
    const hasInput = await run(`(function () {
      var box = document.querySelector('.conhost-input');
      if (!box || box.style.display === 'none') { return false; }
      return !!box.querySelector('input') &&
        box.querySelectorAll('button').length > 0;
    })()`, 'input-area-on');
    check('开启 allow_input 后出现输入区与功能键', hasInput === true);

    await shot(win, 'conhost-input-area.png');

    const typed = 'set /a 111*3';
    const sent = await run(`(function () {
      var box = document.querySelector('.conhost-input input');
      if (!box) { return 'no-box'; }
      box.value = ${JSON.stringify(typed)};
      box.dispatchEvent(new KeyboardEvent('keydown',
        { key: 'Enter', bubbles: true }));
      return 'ok';
    })()`, 'send-input');
    check('在输入框里输入并回车发送', sent === 'ok', String(sent));

    const got = await waitFor(win,
      "(document.querySelector('.conhost-view') || {}).textContent.indexOf('333') >= 0",
      25000, 'wait-333');
    const after = await run(
      "(document.querySelector('.conhost-view') || {}).textContent || ''", 'after');
    console.log('      内容片段：' + String(after).slice(-200).replace(/\s+/g, ' '));
    check('注入的命令真的在目标控制台里执行了（输出出现 333）', got === true);

    /* 具名功能键也走一遍：发一个 Ctrl+C，只要求不报错（它打在提示符上，
       本来就不会有可见输出）。 */
    const ctrlc = await run(`(function () {
      var btns = Array.prototype.slice.call(
        document.querySelectorAll('.conhost-keys button'));
      var b = btns.filter(function (n) { return n.textContent.indexOf('Ctrl+C') >= 0; })[0];
      if (!b) { return 'no-ctrlc-button'; }
      b.click();
      return 'ok';
    })()`, 'ctrl-c');
    check('功能键（Ctrl+C）按钮可用', ctrlc === 'ok', String(ctrlc));

    const status = await run(
      "(document.querySelector('.conhost-status') || {}).textContent || ''");
    check('发送后状态栏给出了反馈',
      String(status).indexOf('发送') >= 0 || String(status).indexOf('已发送') >= 0,
      String(status));
    await shot(win, 'conhost-after-input.png');
  } else {
    console.log('=== 9. 输入注入：本次跳过（CH_INPUT 未开启，服务端是只读配置）===');
  }

  const failed = results.filter(function (r) { return !r.ok; });
  console.log('\n================ 结果 ================');
  console.log('共 ' + results.length + ' 项，失败 ' + failed.length + ' 项');
  if (OUT) {
    try {
      fs.writeFileSync(path.join(OUT, 'results.json'),
        JSON.stringify({ results: results, consoleErrors: consoleErrors }, null, 2));
    } catch (err) {
      console.log('写 results.json 失败：' + err);
    }
  }

  app.exit(failed.length === 0 ? 0 : 1);
}

app.whenReady().then(function () {
  main().catch(function (err) {
    console.log('★ 冒烟测试自身出错：' + (err && err.stack ? err.stack : err));
    app.exit(3);
  });
});

File: tools/smoke_conhost.py
# -*- coding: utf-8 -*-
"""
控制台镜像 · 真实浏览器冒烟测试的驱动
=====================================

    python tools\\smoke_conhost.py

它做四件事：

1. 起一个**无窗口的控制台**作为确定性目标：`cmd.exe /k`，用 CREATE_NO_WINDOW
   启动（★ 不会在你桌面上弹窗），并让它打印一行唯一的标记文本；
2. 用**临时配置**起一个真实服务子进程（只暴露一个临时目录，不碰真实磁盘），
   并显式打开 `conhost.enabled`（默认是关的）；
3. 登录拿会话 Cookie，把它交给 Electron —— 等价于「用户已经在登录页登录过」；
4. 用真实的 Electron（Chromium）打开虚拟桌面，走一遍界面流程：
   开始菜单 → 控制台镜像 → 列表 → 选中目标 → 内容区真的画出字符栅格。

为什么选「无窗口控制台」当目标
------------------------------
  * 它是**确定性**的：pid 与标记文本都是本次现造的，不依赖用户桌面上恰好
    开着什么；
  * 它顺带覆盖了「没有窗口的后台控制台」这条路径（真实桌面上很常见）；
  * CREATE_NO_WINDOW 不会弹窗，跑测试不会打扰正在用机器的人。

★ 它**不进 unittest 套件**：需要 node + Electron，而且要有可用的桌面会话。
  与 tools/smoke_photos.py 一样做成「手动跑的工具」，而不是让整套测试因为它变红。

找不到 Electron 时给一句明确的提示并退出（退出码 2），不是静默跳过。
"""

from __future__ import annotations

import json
import os
import subprocess
import sys
import tempfile
import time

BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
if BASE_DIR not in sys.path:
    sys.path.insert(0, BASE_DIR)

from tests._harness import ServerProcess          # noqa: E402

SMOKE_JS = os.path.join(BASE_DIR, "tools", "smoke_conhost.js")

ADMIN_USER = "tester"
ADMIN_PASSWORD = "test-password-123"

CREATE_NO_WINDOW = 0x08000000


def find_electron():
    """依次找：项目里 desktop-lyrics 装的那份 -> PATH 上的 electron。"""
    candidates = [
        os.path.join(BASE_DIR, "desktop-lyrics", "node_modules", "electron",
                     "dist", "electron.exe"),
        os.path.join(BASE_DIR, "desktop-lyrics", "node_modules", "electron",
                     "dist", "electron"),
    ]
    for path in candidates:
        if os.path.isfile(path):
            return path
    import shutil
    return shutil.which("electron") or shutil.which("electron.exe")


def spawn_target_console():
    """
    起一个无窗口的控制台，让它打印标记并保持存活。

    返回 (进程, 标记文本)。

    ★ 三个都不能改的细节（都实测踩过）：

      1. **三个标准句柄都不能传**（不要 stdin/stdout/stderr 参数）。
         传了 stdout=DEVNULL，`cmd` 的输出就进了 NUL 而不是**它自己的控制台**，
         于是控制台缓冲区里一行都没有 —— 表现是「读取成功但 0 行」，
         而状态栏还是正常显示尺寸，非常像前端渲染坏了。
         不传句柄时 Windows 会把它们指向子进程新建的那个控制台，
         输出才会真的落进缓冲区（本功能读的就是那个缓冲区）。
         ★ 这一条同时也是功能的现实边界：真实桌面上那些
         `cmd /c "xxx.bat"` 包装进程往往把输出重定向走了，它们的控制台
         本来就是空的 —— 界面里看到空内容不一定是 bug。

      2. **stdin 绝不能接 DEVNULL。** `cmd /k` 从 stdin 读到 EOF 会立刻退出，
         于是我们拿到的是一个**已经死掉的 pid**；而 AttachConsole 对已退出的
         pid 返回的是 ERROR_ACCESS_DENIED(5)，和「权限不足」一模一样，
         能把人往错误方向带很久。

      3. 用 `CREATE_NO_WINDOW` 而不是 `CREATE_NEW_CONSOLE`：前者不弹窗，
         不会打扰正在用机器的人；而且**无窗口的控制台照样能被附加与读取**
         （实测证实），顺带覆盖了这条路径。
    """
    mark = "CONHOST-SMOKE-%d" % (os.getpid() * 1000 + int(time.time()) % 1000)
    proc = subprocess.Popen(
        ["cmd.exe", "/k", "echo %s" % mark],
        creationflags=CREATE_NO_WINDOW)
    # ↑ 刻意不传 stdin/stdout/stderr，见上面第 1、2 条
    return proc, mark


def main() -> int:
    if not sys.platform.startswith("win"):
        print("★ 控制台镜像只在 Windows 上有意义。")
        return 2

    # --input：额外打开 conhost.allow_input 并把「输入注入」那一组检查也跑掉。
    # 默认不跑，因为服务端默认就是只读的 —— 两种配置各跑一次才是完整覆盖。
    want_input = "--input" in sys.argv[1:]

    electron = find_electron()
    if not electron:
        print("★ 没有找到 Electron，无法做浏览器冒烟测试。")
        print("  装一个即可：cd desktop-lyrics && npm install（或把 electron 放进 PATH）")
        return 2

    work = tempfile.mkdtemp(prefix="fw-conhost-smoke-")
    root = os.path.join(work, "root")
    out_dir = os.path.join(work, "shots")
    os.makedirs(root, exist_ok=True)
    os.makedirs(out_dir, exist_ok=True)

    server = None
    target = None
    try:
        target, mark = spawn_target_console()
        time.sleep(1.2)
        print("目标控制台：pid=%d（无窗口），标记=%s" % (target.pid, mark))

        def extra(cfg):
            cfg["conhost"]["enabled"] = True
            # 只读那一轮把 allow_input 关掉（这是默认姿态，必须能纯只读）；
            # 带 --input 的那一轮才打开，用来验证注入链路。
            cfg["conhost"]["allow_input"] = bool(want_input)

        server = ServerProcess(
            [{"id": "r1", "name": "root", "path": root}],
            username=ADMIN_USER, password=ADMIN_PASSWORD,
            extra_config=extra).start()

        client = server.login_client()

        # 先确认接口这一侧真的能看见那个目标 —— 浏览器那边要是找不到，
        # 就能立刻区分「接口没枚举到」还是「前端没渲染」。
        status, data = client.json("GET", "/api/conhost/list")
        if status != 200:
            print("★ 列举控制台失败：%r %r" % (status, data))
            return 2
        pids = []
        for item in data.get("items") or []:
            pids.extend(m["pid"] for m in item.get("members") or [])
        print("接口枚举到 %d 个控制台；目标 pid 在里面：%s"
              % (data.get("count", 0), target.pid in pids))
        if target.pid not in pids:
            print("★ 目标控制台没有被枚举到，浏览器那边必然找不到。"
                  "先查 conhost.list_consoles 的聚合与过滤。")
            return 2

        env = dict(os.environ)
        env["CH_PORT"] = str(server.port)
        env["CH_COOKIE"] = client.cookie
        env["CH_OUT"] = out_dir
        env["CH_PID"] = str(target.pid)
        env["CH_MARK"] = mark
        env["CH_INPUT"] = "1" if want_input else "0"

        print("用真实 Electron 打开虚拟桌面（Chromium）…\n")
        proc = subprocess.run(
            [electron, SMOKE_JS, "--no-sandbox", "--disable-gpu"],
            cwd=BASE_DIR, env=env, timeout=420,
            stdout=subprocess.PIPE, stderr=subprocess.STDOUT)

        print((proc.stdout or b"").decode("utf-8", "replace"))

        results_path = os.path.join(out_dir, "results.json")
        if os.path.isfile(results_path):
            with open(results_path, encoding="utf-8") as fh:
                payload = json.load(fh)
            failed = [r for r in payload.get("results", []) if not r.get("ok")]
            print("截图与结果已写到：%s" % out_dir)
            for item in failed:
                print("  !! %s —— %s" % (item["label"], item["extra"]))
        return proc.returncode
    finally:
        if target is not None and target.poll() is None:
            subprocess.run(["taskkill", "/PID", str(target.pid), "/T", "/F"],
                           capture_output=True)
        if server is not None:
            server.stop()
            server.cleanup()
        print("工作目录（含截图）：%s" % work)


if __name__ == "__main__":
    sys.exit(main())
